benchmark/Streamly/Benchmark/FileSystem/DirIO.hs

```diff
@@ -0,0 +1,153 @@
+import qualified Streamly.Data.Unfold as Unfold
+import qualified Streamly.Internal.Data.Unfold as Unfold
+    (either, nil)
+import qualified Streamly.Internal.FileSystem.DirIO as Dir
+    (readEitherChunks, readEitherPaths, eitherReaderPaths)
+import qualified Streamly.FileSystem.Path as Path
+import qualified Streamly.Internal.FileSystem.Path as Path (toChunk)
+#if !defined(mingw32_HOST_OS) && !defined(__MINGW32__)
+import qualified Streamly.Internal.FileSystem.Posix.ReadDir as Dir
+    (readEitherByteChunks)
+#endif
+
+import Prelude hiding (last, length)
+import Test.Tasty.Bench hiding (env)
+
+--------------------------------------------------------------------------------
+-- Functions
+--------------------------------------------------------------------------------
+
+_concatIterateWith :: (StreamK.StreamK IO (Either Path Path)
+                      -> StreamK.StreamK IO (Either Path Path)
+                      -> StreamK.StreamK IO (Either Path Path))
+                      -> Stream IO (Either Path Path) -> Stream IO (Either Path Path)
+_concatIterateWith f =
+      StreamK.toStream
+    . StreamK.concatIterateWith f (StreamK.fromStream . streamDir)
+    . StreamK.fromStream
+
+_mergeIterateWith :: (StreamK.StreamK IO (Either Path Path)
+                      -> StreamK.StreamK IO (Either Path Path)
+                      -> StreamK.StreamK IO (Either Path Path))
+                      -> Stream IO (Either Path Path) -> Stream IO (Either Path Path)
+_mergeIterateWith f =
+      StreamK.toStream
+    . StreamK.mergeIterateWith f (StreamK.fromStream . streamDir)
+    . StreamK.fromStream
+
+streamDir :: Either Path b -> Stream IO (Either Path Path)
+streamDir = either Dir.readEitherPaths (const Stream.nil)
+
+unfoldDir :: Unfold IO (Either Path b) (Either Path Path)
+unfoldDir = Unfold.either Dir.eitherReaderPaths Unfold.nil
+
+_streamDirMaybe :: Either Path b -> Maybe (Stream IO (Either Path Path))
+_streamDirMaybe = either (Just . Dir.readEitherPaths) (const Nothing)
+
+_streamDirByteChunked
+    :: Either [Path] b -> Stream IO (Either [Path] (Array Word8))
+_streamDirByteChunked = either Dir.readEitherByteChunks (const Stream.nil)
+
+_streamDirByteChunkedMaybe
+    :: Either [Path] b -> Maybe (Stream IO (Either [Path] (Array Word8)))
+_streamDirByteChunkedMaybe =
+    either (Just . Dir.readEitherByteChunks) (const Nothing)
+
+_streamDirChunked :: Either [Path] b -> Stream IO (Either [Path] [Path])
+_streamDirChunked = either Dir.readEitherChunks (const Stream.nil)
+
+streamDirChunkedMaybe :: Either [Path] b -> Maybe (Stream IO (Either [Path] [Path]))
+streamDirChunkedMaybe = either (Just . Dir.readEitherChunks) (const Nothing)
+
+#if !defined(mingw32_HOST_OS) && !defined(__MINGW32__)
+-- Fastest implementation, only works for posix as of now.
+listDirByteChunked :: FilePath -> IO ()
+listDirByteChunked inp = do
+    Stream.fold Fold.drain
+        $ Stream.catRights
+        $ Stream.concatIterateDfs streamDirChunkedMaybe
```

Comment:
The fastest version uses the concatIterateWith StreamK.append as in the streamly-examples code. Same in the other two benchmarks as well. We can have multiple variants of the benchmarks using concateIterateDfs/Bfs etc as well.